aspnetcore/fundamentals/url-rewriting.md

```diff
@@ -0,0 +1,303 @@
+
+[!code-csharp[Main](url-rewriting/sample/Startup.cs?name=snippet1&highlight=2)]
+
+In a browser with developer tools enabled, make a request to the sample application with the URL **http://localhost:5000/redirect-rule/1234/5678**. The regex matches the request path on `redirect-rule/(.*)`, and the path is replaced with **redirected/1234/5678**. The redirect URL is sent back to the client with a 302 (Found) status code. The browser makes a new request at the redirect URL, which will appear in the browser's address bar. Since no rules in the sample application match on the redirect URL, the second request receives a 200 (OK) response from the application and the body of the response shows the redirect URL. A complete roundtrip is made to the server when a URL is *redirected*.
+
+Original Request: **http://localhost:5000/redirect-rule/1234/5678**
+
+![Browser window with Developer Tools tracking the requests and responses](url-rewriting/_static/add_redirect.png)
+
+The part of the expression contained by parentheses is called a *capture group*. The dot (`.`) of the expression means *match any character*, and the asterisk (`*`) signifies to *match the preceding character zero or more times*. Therefore, the last two path segments of the URL, `1234/5678`, are captured by capture group `(.*)`. Any value you provide in the request URL after **redirect-rule/** will be captured by this single capture group.
+
+In the replacement string, captured groups are injected into the string with the dollar sign (`$`) followed by the sequence number of the capture. The first capture group value is obtained with `$1`, the second with `$2`, and they continue in sequence for the capture groups in your regex. There is only one captured group in the redirect rule regex in the sample application, so there is only one injected group in the replacement string, which is `$1`. When the rule is applied, the URL becomes **http://localhost:5000/redirected/1234/5678**.
+
+#### URL Redirect to a secure endpoint (permanent)
+Use `AddRedirectToHttpsPermanent()` to redirect insecure requests to the same host and path with secure HTTPS protocol (**https://** on port 465). The middleware will set the status code to 301 (Moved Permanently).
+```csharp
+var options = new RewriteOptions()
+    .AddRedirectToHttpsPermanent();
+
+app.UseRewriter(options);
+```
+The sample application is capable of demonstrating how to use `AddRedirectToHttpsPermanent()`. Add the extension method to the `RewriteOptions()`. Make an insecure request to the application at any URL. In order to see the response in a browser, you will probably need to dismiss a browser security warning that the self-signed certificate is untrusted.
+
+Original Request: **http://localhost:5000/secure**
+
+![Browser window with Developer Tools tracking the requests and responses](url-rewriting/_static/add_redirect_to_https_permanent.png)
+
+#### URL Redirect to a secure endpoint
```

Comment:
... well ... that would break my doc structure. I have one section per feature, and I think it's more natural to show the limited case (permanent 443) before showing that those are settable with the next feature.

> is only a minor variant

In code, yes ... but in explaining things in a document, I'm not so sure. It sets the status code and port, while the other doesn't. That all has to be explained out. We have to show the code anyway, because in the sample they must wire those bits up if they want to play with them.

I see combining here as causing more problems that it might solve. I'd prefer not to combine unless they :boy: :gun: tell me to. :smile:

@Rick-Anderson @tdykstra What do you want to do here?